java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/HashTableBasedDictionaryBuilder.java

```diff
@@ -0,0 +1,218 @@
+ * This class builds the dictionary based on a hash table.
+ * Each add operation can be finished in O(1) time,
+ * where n is the current dictionary size.
+ * </p>
+ * <p>
+ * The dictionary builder is intended to build a single dictionary.
+ * So it cannot be used for different dictionaries.
+ * </p>
+ * <p>Below gives the sample code for using the dictionary builder
+ * <pre>{@code
+ * ArrowBufHasherDictionaryBuilder dictionaryBuilder = ...
+ * dictionaryBuilder.startBuild();
+ * ...
+ * dictionaryBuild.addValue(newValue);
+ * ...
+ * dictionaryBuilder.endBuild();
+ * }</pre>
+ * </p>
+ * <p>
+ *   With the above code, the dictionary vector will be populated,
+ *   and it can be retrieved by the {@link HashTableBasedDictionaryBuilder#getDictionary()} method.
+ *   After that, dictionary encoding can proceed with the populated dictionary encoder.
+ * </p>
+ *
+ * @param <V> the dictionary vector type.
+ */
+public class HashTableBasedDictionaryBuilder<V extends ElementAddressableVector> {
+
+  /**
+   * The dictionary to be built.
+   */
+  private final V dictionary;
+
+  /**
+   * The total number of data items probed.
+   */
+  private int totalValueCount;
+
+  /**
+   * The number of items in the dictionary.
+   */
+  private int distinctValueCount;
+
+  /**
+   * If null should be encoded.
+   */
+  private final boolean encodeNull;
+
+  /**
+   * The hash map for distinct dictionary entries.
+   * The key is the pointer to the dictionary element, whereas the value is the index in the dictionary.
+   */
+  private HashMap<ArrowBufPointer, Integer> hashMap = new HashMap<>();
+
+  /**
+   * A flag indicating if dictionary building is finished.
+   */
+  private boolean buildFinished = false;
+
+  /**
+   * The hasher used for calculating the hash code.
+   */
+  private final ArrowBufHasher hasher;
+
+  /**
+   * Next pointer to try to add to the hash table.
+   */
+  private ArrowBufPointer nextPointer;
+
+  /**
+   * Constructs a hash table based dictionary builder.
+   *
+   * @param dictionary the dictionary to populate.
+   */
+  public HashTableBasedDictionaryBuilder(V dictionary) {
+    this(dictionary, false);
+  }
+
+  /**
+   * Constructs a hash table based dictionary builder.
+   *
+   * @param dictionary the dictionary to populate.
+   * @param encodeNull if null should be encoded.
+   */
+  public HashTableBasedDictionaryBuilder(V dictionary, boolean encodeNull) {
+    this(dictionary, encodeNull, SimpleHasher.INSTANCE);
+  }
+
+  /**
+   * Constructs a hash table based dictionary builder.
+   *
+   * @param dictionary the dictionary to populate.
+   * @param encodeNull if null should be encoded.
+   * @param hasher     the hasher used to compute the hash code.
+   */
+  public HashTableBasedDictionaryBuilder(V dictionary, boolean encodeNull, ArrowBufHasher hasher) {
+    this.dictionary = dictionary;
+    this.encodeNull = encodeNull;
+    this.hasher = hasher;
+    this.nextPointer = new ArrowBufPointer(hasher);
+  }
+
+  /**
```

Comment:
is it important to keep the state of startBuild and endBuild.  Is seems like we can set ValueCount  whenever getDictionary is called if users want an incremental dictionary?

Reply:
Revised. Thanks for your suggestion.